Agent Registry Foundation

muthukumaranR opened the pull request:
## Summary

Adds Agent Registry system for auto-discovery and configuration of AKD agents.

## Key Features

- **Auto-discovery**: Automatically finds and registers agents from known modules
- **JSON persistence**: Generates `akd/mapping/agent_registry.json` with agent definitions
- **Selective loading**: Configure which agents to load via settings
- **Schema extraction**: Captures input/output schemas from agent classes

## Generate Registry JSON

The registry auto-generates when the JSON file is missing or empty:

```python
from akd.planner.registry import AgentRegistry

# Auto-discovers all available agents
registry = AgentRegistry()

# Or configure specific agents only
from akd.planner.config import AgentRegistryConfig
config = AgentRegistryConfig(enabled_agents=["query", "extraction"])
registry = AgentRegistry(config)
```

## Configuration

Configure via `AgentRegistryConfig`:

```python
config = AgentRegistryConfig(
    registry_path="custom/path/registry.json",  # Default: akd/mapping/agent_registry.json
    auto_discover=True,                         # Auto-find agents when missing
    enabled_agents=["query", "extraction"],    # Limit to specific agents ([] = all)
    validate_schemas=True                       # Ensure agents have valid schemas
)
```

## How It Works

1. **Discovery**: Scans `KNOWN_AGENTS` list for agent modules
2. **Schema extraction**: Uses `model_json_schema()` to capture agent input/output definitions
3. **JSON generation**: Creates structured registry file with agent metadata
4. **Loading**: Provides methods to query agents by ID, tag, or enabled status

The generated JSON contains agent definitions with schemas, descriptions, tags.

muthukumaranR committed:
Add Agent Registry Foundation

Implements auto-discovery and configuration management for AKD agents:
- Registry auto-discovers agents from known modules
- Generates agent_registry.json with schema definitions
- Supports selective agent loading via configuration
- Framework-compliant with proper type annotations

NISH1001 commented on akd/planner/registry.py:
```diff
@@ -0,0 +1,286 @@
+class FieldDefinition(BaseModel):
+    """Individual field definition for agent schemas."""
+    
+    name: str = Field(description="Field name")
+    type: str = Field(description="Field type")
```
Are these required fields? Then we could use elipses `...` in the `Field(..., descriptoin="...")`

muthukumaranR replied:
thanks for catching this - updated.

muthukumaranR replied:
[822ce86](https://github.com/NASA-IMPACT/accelerated-discovery/pull/198/commits/822ce86c68bc6e47afcc48c9ecd152583a14de9f)

NISH1001 commented on akd/planner/registry.py:
```diff
@@ -0,0 +1,286 @@
+    type: str = Field(description="Field type")
+    description: str = Field(description="Field description")
+    required: bool = Field(default=True, description="Whether field is required")
+    default: Union[str, int, float, bool, List[Any], None] = Field(default=None, description="Default value if any")
```
Nitpick:
Maybe let's start to follow new python 3.12+ typing style with 
```
default: str | int | float | bool | list[Any] | None = ...
```

muthukumaranR replied:
changed to use the python 3.12 types here: [822ce86](https://github.com/NASA-IMPACT/accelerated-discovery/pull/198/commits/822ce86c68bc6e47afcc48c9ecd152583a14de9f)

muthukumaranR replied:
updated to use 3.12 python types

NISH1001 commented on akd/planner/config.py:
```diff
@@ -0,0 +1,32 @@
+    """Configuration for the Agent Registry."""
+    
+    registry_path: str = Field(
+        default="akd/mapping/agent_registry.json",
```
use probably adk root-based path

```python
from akd.utils import get_akd_root
```
and take it from there? it iwll be better...We've done it on several places. [eg](https://github.com/NASA-IMPACT/accelerated-discovery/blob/develop/akd/tools/code_search.py#L251)

muthukumaranR replied:
fixed here [d71e6d2](https://github.com/NASA-IMPACT/accelerated-discovery/pull/198/commits/d71e6d201f70b93db46ce597c7f32cb2fbaf6961)

NISH1001 commented on akd/planner/registry.py:
```diff
@@ -0,0 +1,286 @@
+        ("gap_analysis", "akd.agents.gap_analysis.gap_analysis", "GapAgent"),
+        ("storm", "akd.agents.storm.storm", "StormAgent"),
+        ("aspect_search", "akd.agents.search.aspect_search.aspect_search", "AspectSearchAgent"),
+    ]
```
Should we do this as mapping of Dictionary to tuple?

```python
KNOWN_AGENTs  = {
"query" : (akd.agents.query, QueryAgent)
}
```

etc?

muthukumaranR replied:
734ac77

NISH1001 commented on akd/planner/registry.py:
```diff
@@ -0,0 +1,286 @@
+    global _registry
+    if _registry is None:
+        _registry = AgentRegistry(config)
+    return _registry
```
This exhibits singleton, right? or maybe we can inherently make the registry singleton?

```python
class AgentRegistry:

_registry = None

def __new__(self....)->
... if _regsitry is None: intialize new
else just use _registry?
```

what do you think?

Gemini-recommended snippet:
```python
class AgentRegistry:
    # 1. A class attribute to hold the single instance
    _instance = None

    # 2. Override the __new__ method
    def __new__(cls, *args, **kwargs):
        # If the single instance doesn't exist yet...
        if cls._instance is None:
            # ...create it by calling the parent class's __new__ method
            # and store it in our class attribute.
            cls._instance = super().__new__(cls)
        # Always return the single instance
        return cls._instance

    # The rest of the class ( __init__, _load_or_discover, etc.) stays the same
    def __init__(self, config: Optional[AgentRegistryConfig] = None):
        """Initialize the agent registry."""
        # Note: __init__ will run every time you call AgentRegistry(),
        # so you might need to add a check to prevent re-initialization.
        # For example: if hasattr(self, 'config'): return
        self.config = config or AgentRegistryConfig()
        self.registry_data: AgentRegistryData = AgentRegistryData()
        self._load_or_discover()

    # ... other methods ...
```

This way we can always reuse the same object and minimizes potential side-effects of registry mutation. IF mutation / registry is needed to be updated at runtime, we can allow through some methods and it will be scoped to everywhere since only single instance can be created/reused.

muthukumaranR replied:
good suggestion - [99cc1c2](https://github.com/NASA-IMPACT/accelerated-discovery/pull/198/commits/99cc1c2a7367543adf7ee05d5831422593e5ea01)

  Replaced the global _registry variable with singleton implementation:
  - Added _instance and _initialized class attributes
  - Implemented __new__ to return single instance
  - Modified __init__ to initialize only once

NISH1001 reviewed (commented):
singleton comments

muthukumaranR committed:
Refactor AgentRegistry to use proper singleton pattern

- Replace global _registry variable with __new__ singleton implementation
- Add _instance and _initialized class attributes for singleton management
- Update __init__ to initialize only once using _initialized flag
- Simplify get_agent_registry() to directly return AgentRegistry(config)
- Update tests to reset singleton state instead of global variable

muthukumaranR committed:
Make FieldDefinition required fields explicit

Use ellipses (...) in Field definitions for name, type, and description
to clearly indicate these are required fields without defaults.

muthukumaranR committed:
Modernize to Python 3.12+ typing syntax

Replace typing imports with modern syntax and remove unused imports.

muthukumaranR committed:
Refactor KNOWN_AGENTS to dictionary mapping

Change from list of tuples to dict mapping agent_id to (module_path, class_name).
Update _discover_agents loop to use .items() iteration.

muthukumaranR committed:
Use get_akd_root() for registry path resolution

Replace hardcoded path with get_akd_root() / "akd" / "mapping" / "agent_registry.json"
to ensure proper path resolution regardless of working directory.

Also modernize List[str] to list[str] in typing.

NISH1001 commented on akd/planner/registry.py:
```diff
@@ -67,29 +68,43 @@ class AgentRegistry:
+            self.config = config or AgentRegistryConfig()
+            self.registry_data: AgentRegistryData = AgentRegistryData()
+            self._load_or_discover()
+            AgentRegistry._initialized = True
```
Nice. This makes it cleaner to put singleton.

muthukumaranR committed:
Fix registry test failures for CI/CD compatibility

- Fix path assertion to use endswith() for environment portability
- Fix selective agent loading test with proper module-specific mocking

muthukumaranR committed:
check for directory exists==True

muthukumaranR committed:
add a classmethod to reset the singleton instance (will be only used for tests)

muthukumaranR committed:
reset the registry for testing
